File: 03.CPP/0076.cpp
#include <iostream>
#include <unordered_map>
#include <string>
#include <climits>

class Solution {
public:
    std::string minWindow(std::string s, std::string t) {
        std::unordered_map<char, int> need;

        for (char c : t) {
            need[c]++;
        }

        int required = t.length();
        int left = 0;
        int minLen = INT_MAX;
        int start = 0;

        for (int right = 0; right < s.length(); right++) {

            if (need[s[right]] > 0) {
                required--;
            }

            need[s[right]]--;

            while (required == 0) {

                if (right - left + 1 < minLen) {
                    minLen = right - left + 1;
                    start = left;
                }

                need[s[left]]++;

                if (need[s[left]] > 0) {
                    required++;
                }

                left++;
            }

        }

        if (minLen == INT_MAX) {
            return "";
        }

        return s.substr(start, minLen);
    }
};

int main() {
    Solution sol;

    std::string s = "ADOBECODEBANC";
    std::string t = "ABC";

    std::string result = sol.minWindow(s, t);
    std::cout << "\nThe correct substring is: " << result << "\n";

    return 0;
}

File: 01.Python/0076.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        need = {}

        for char in t:
            need[char] = need.get(char, 0) + 1

        required = len(t)

        left = 0
        minLen = float('inf')
        start = 0

        for right in range(len(s)):
            if (need.get(s[right], 0) > 0):
                required -= 1
            
            need[s[right]] = need.get(s[right], 0) - 1

            while required == 0:

                if right - left + 1 < minLen:
                    minLen = right - left + 1
                    start = left
                
                need[s[left]] = need.get(s[left], 0) + 1

                if need[s[left]] > 0:
                    required += 1
                
                left += 1

        if minLen == float('inf'):
            return ""

        return s[start:start + minLen]

sol = Solution()

s = "ADOBECODEBANC"
t = "ABC"

result = sol.minWindow(s, t)
print(result)